WordPress/gutenberg
# Navigation: Try adding navigation link variants via server

The navigation block currently allows insertion of Links, Post Links, Page Links, Tag Links and Category Links. This PR adds the ability to insert links for custom post types and taxonomies. If we install a Portfolio plugin for example, we should be able to add Portfolio Links.

In master, the list of link variations is hardcoded. This PR experiments with registering navigation link variants using `register_block_type_from_metadata` and passing not only a render callback, but the variations array. One benefit of doing this is being able to fetch post types and taxonomies and filter by `show_in_nav_menus` which is not exposed via the REST API. It also has some flexibility in pulling other data from the server without needing to bloat the API response (values here are unlikely to be reused elsewhere).

Part of https://github.com/WordPress/gutenberg/issues/24814, Alternative to https://github.com/WordPress/gutenberg/pull/29011, Requires https://github.com/WordPress/wordpress-develop/pull/1015 to test.

| Before  | After |
| ------------- | ------------- |
| <img width="1206" alt="Screen Shot 2021-02-24 at 8 40 12 PM" src="https://user-images.githubusercontent.com/1270189/109104300-976a2580-76e0-11eb-8d32-2f50cff7bd36.png"> | <img width="1397" alt="Screen Shot 2021-03-01 at 12 47 07 PM" src="https://user-images.githubusercontent.com/1270189/109556977-567d6280-7a8c-11eb-8248-e7e50824f859.png"> |

https://user-images.githubusercontent.com/1270189/109557915-955fe800-7a8d-11eb-9e97-7d6bad835cd9.mp4

### TODO
- [x] ~~The server definition is being blown away by the local block.json.~~ Drop placeholder change in favor of https://github.com/WordPress/gutenberg/pull/29213
- [x] Rebase to pull in https://github.com/WordPress/gutenberg/pull/29302
- [x] Move custom post icon to it's own PR https://github.com/WordPress/gutenberg/pull/29232
- [x] Add new post_type_labels in another patch
- [x] Update visibility of links other CPTs are drafts. Add more tests
- [x] Think through steps of safely disabling this if core changes aren't available. We currently support WP5.5+
- [x] WIP I still need to polish the wordpress-dev PR (https://github.com/WordPress/wordpress-develop/pull/1015) and add additional test cases 
- [x] Update screenshots on summary again

### Testing Instructions with https://github.com/WordPress/wordpress-develop/pull/1015
- Setup a local environment of your choice using [wordpress-develop](https://github.com/WordPress/wordpress-develop)
- Checkout the wordpress develop banch in https://github.com/WordPress/wordpress-develop/pull/1015.
- Setup gutenberg and checkout `try/block-server-variants` . One way of doing this on a local install is to link this via `ln -s your-gutenberg-checkout-path your-wordpress-local-path/wp-content/plugins/gutenberg`. Run `npm run build` in your gutenberg checkout
- Install a plugin with some custom post types and taxonomies
- Edit a post
- Click on the block inserter, and add a navigation block
- Click start with empty
- Click on the block inserter in navigation
- Click on show more. We should see installed custom post types and taxonomies as link variant options
- Verify that there are no errors in console or in the server error log.

| With Server Patch |
|------|
| <img width="1206" alt="109104191-6853b400-76e0-11eb-89b4-320156250f73" src="https://user-images.githubusercontent.com/1270189/109329827-e8c50280-780f-11eb-8e11-38383bf9eac5.png"> |

#### If Custom Post Type or Taxonomy does not define new `item_link` or `item_link_description` labels it shows the fallback:

<img width="1397" alt="Screen Shot 2021-03-01 at 12 34 45 PM" src="https://user-images.githubusercontent.com/1270189/109555968-0d78de80-7a8b-11eb-9679-2d37d1a21b8b.png">

#### With `item_link` or `item_link_description` labels 

<img width="1397" alt="Screen Shot 2021-03-01 at 12 47 07 PM" src="https://user-images.githubusercontent.com/1270189/109557114-83ca1080-7a8c-11eb-868c-70b66da38589.png">

This will require manually modifying a taxonomy or post_type registration since no plugin has added this yet:
<img width="944" alt="Screen Shot 2021-03-01 at 12 50 22 PM" src="https://user-images.githubusercontent.com/1270189/109557321-cbe93300-7a8c-11eb-87d5-e7c169b88581.png">


#### CPT Drafts should not render on Navigation Block
- Publish a few custom post type posts
- Add the CPT links to a navigation block
- Verify that the links render on the frontend
- Set one of the CPT posts to 'draft' status, by updating visibility settings to `private`
- The private CPT post link should not render on the published post or page view
<img width="381" alt="Screen Shot 2021-02-26 at 10 17 35 AM" src="https://user-images.githubusercontent.com/1270189/109339011-dea90100-781b-11eb-86c7-c0592bcd8fb8.png">

### Testing Instructions for other WP versions
- Checkout this branch
- Install a plugin with some custom post types and taxonomies
- Edit a post
- Click on the block inserter, and add a navigation block
- Click start with empty
- Click on the block inserter in navigation
- We should see the fallback variations of (Link, Post, Page, Tag, Category) but not the installed variations.
- Verify that there are no errors in console or in the server error log.

| Fallback |
|-----|
|  <img width="693" alt="Screen Shot 2021-02-26 at 8 49 47 AM" src="https://user-images.githubusercontent.com/1270189/109329686-be734500-780f-11eb-8e00-f3e63160d1ed.png"> |
